pkg/services/ngalert/notifier/channels/discord.go

```diff
@@ -0,0 +1,125 @@
+package channels
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+
+	gokit_log "github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/alertmanager/types"
+
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/alerting"
+	old_notifiers "github.com/grafana/grafana/pkg/services/alerting/notifiers"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type DiscordNotifier struct {
+	old_notifiers.NotifierBase
+	log         log.Logger
+	tmpl        *template.Template
+	Description string
+	Content     string
+	WebhookURL  string
+}
+
+func NewDiscordNotifier(model *NotificationChannelConfig, t *template.Template) (*DiscordNotifier, error) {
+	if model.Settings == nil {
+		return nil, alerting.ValidationError{Reason: "No Settings Supplied"}
+	}
+
+	discordURL := model.Settings.Get("url").MustString()
+	if discordURL == "" {
+		return nil, alerting.ValidationError{Reason: "Could not find webhook url property in settings"}
+	}
+
+	content := model.Settings.Get("message").MustString(`{{ template "default.message" . }}`)
+
+	return &DiscordNotifier{
+		NotifierBase: old_notifiers.NewNotifierBase(&models.AlertNotification{
+			Uid:                   model.UID,
+			Name:                  model.Name,
+			Type:                  model.Type,
+			DisableResolveMessage: model.DisableResolveMessage,
+			Settings:              model.Settings,
+			SecureSettings:        model.SecureSettings,
+		}),
+		Content:    content,
+		WebhookURL: discordURL,
+		log:        log.New("alerting.notifier.discord"),
+		tmpl:       t,
+	}, nil
+}
+
+func (d DiscordNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	data := notify.GetTemplateData(ctx, d.tmpl, as, gokit_log.NewNopLogger())
+	alerts := types.Alerts(as...)
+
+	bodyJSON := simplejson.New()
+	bodyJSON.Set("username", "Grafana")
+
+	var tmplErr error
+	tmpl := notify.TmplText(d.tmpl, data, &tmplErr)
+	if d.Content != "" {
+		bodyJSON.Set("content", tmpl(d.Content))
+	}
+	if tmplErr != nil {
+		return false, errors.Wrap(tmplErr, "failed to template discord message")
+	}
```

Comment:
This should be checked at the end _after_ having called all the `tmpl(...)`. This variable is populated if there was any error in `tmpl()` calls.